### PR TITLE
Add all-in-one CI testing

### PR DESCRIPTION
**Pull request by markgoddard**
* Add [kayobe-automation](https://github.com/stackhpc/kayobe-automation) submodule & configuration
* Add kayobe to `requirements.txt`
* Add an all-in-one Terraform configuration & README
* Add a playbook to create test resources
* Populate kayobe-config Tempest configuration for refstack
* Add Github actions workflows to build the kayobe image, and run all-in-one tests againsts PRs

### CODE REVIEW

**Comment by markgoddard** [comment]
It probably needs a little more work, but wanted to get your thoughts @jovial

**Review by jovial — APPROVED** [review]
Looks good to me.

**Comment by markgoddard** [comment]
Is there anything in here that would conflict with existing consumer configuration? Perhaps `automation.conf/config.sh`?

**Comment by jovial** [comment]
> Is there anything in here that would conflict with existing consumer configuration? Perhaps `automation.conf/config.sh`?

I've been making things conditional on the kayobe environment:

```
  if [ "$KAYOBE_ENVIRONMENT" == "aio" ]; then
    # Seem to get servers failing to spawn with higher concurrency
    export TEMPEST_CONCURRENCY=1
  fi
```

so we could supply a basic configuration that could be used for client side CI and put any upstream specifics behind guards like this. Yes it would merge conflict, but at least that provides a nice way of keeping it up to date. What do you reckon? The one slight gotcha is that you need to supply KAYOBE_ENVIRONMENT instead of relying on `.environment` as `kayobe-env` will not have been loaded yet. That could possibly be resolved if we just read that file as well.

Edit: reworded slightly

**Comment by markgoddard** [comment]
> > Is there anything in here that would conflict with existing consumer configuration? Perhaps `automation.conf/config.sh`?
> 
> I've been making things conditional on the kayobe environment:
> 
> ```
>   if [ "$KAYOBE_ENVIRONMENT" == "aio" ]; then
>     # Seem to get servers failing to spawn with higher concurrency
>     export TEMPEST_CONCURRENCY=1
>   fi
> ```
> 
> so we could supply a basic configuration that could be used for client side CI and put any upstream specifics behind guards like this. Yes it would merge conflict, but at least that provides a nice way of keeping it up to date. What do you reckon? The one slight gotcha is that you need to supply KAYOBE_ENVIRONMENT instead of relying on `.environment` as `kayobe-env` will not have been loaded yet. That could possibly be resolved if we just read that file as well.
> 
> Edit: reworded slightly

I did that for the concurrency (adapted from SOS), is there anything else it needs to be done for?

**Comment by jovial** [comment]
> > > Is there anything in here that would conflict with existing consumer configuration? Perhaps `automation.conf/config.sh`?
> > 
> > 
> > I've been making things conditional on the kayobe environment:
> > ```
> >   if [ "$KAYOBE_ENVIRONMENT" == "aio" ]; then
> >     # Seem to get servers failing to spawn with higher concurrency
> >     export TEMPEST_CONCURRENCY=1
> >   fi
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > so we could supply a basic configuration that could be used for client side CI and put any upstream specifics behind guards like this. Yes it would merge conflict, but at least that provides a nice way of keeping it up to date. What do you reckon? The one slight gotcha is that you need to supply KAYOBE_ENVIRONMENT instead of relying on `.environment` as `kayobe-env` will not have been loaded yet. That could possibly be resolved if we just read that file as well.
> > Edit: reworded slightly
> 
> I did that for the concurrency (adapted from SOS), is there anything else it needs to be done for?

Don't have any good examples really, but I've started shipping some additional test lists:

```
KAYOBE_AUTOMATION_TEMPEST_CONF_OVERRIDES="${KAYOBE_AUTOMATION_CONFIG_PATH}/tempest/tempest.overrides.conf"

if [ ! -z ${KAYOBE_ENVIRONMENT:+x} ]; then
  KAYOBE_AUTOMATION_TEMPEST_CONF_OVERRIDES="${KAYOBE_AUTOMATION_CONFIG_PATH}/tempest/tempest-${KAYOBE_ENVIRONMENT}-${KAYOBE_AUTOMATION_TEMPEST_LOADLIST:-}.overrides.conf"

  # Check if loadlist specific overrides exist, if not fallback to environment overrides.
  if [ ! -e "${KAYOBE_AUTOMATION_TEMPEST_CONF_OVERRIDES}" ]; then
      KAYOBE_AUTOMATION_TEMPEST_CONF_OVERRIDES="${KAYOBE_AUTOMATION_CONFIG_PATH}/tempest/tempest-${KAYOBE_ENVIRONMENT}.overrides.conf"
  fi
fi
```
Trying to establish a convention for where to place tempest overrides. It basically does first found on:

- overrides for specific test list using a specific environment
- overrides for specific environment
- shared overrides

Trying to go for convention over configuration here, so that people don't start doing it in multiple different ways.

Maybe that shouldn't be in the config file, but i started there whilst experimenting.

File: .automation.conf/config.sh [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+#export TEMPEST_PATTERN="${TEMPEST_PATTERN:-tempest.api.compute.servers.test_create_server.ServersTestJSON.test_host_name_is_same_as_server_name}"
+
+KAYOBE_AUTOMATION_TEMPEST_CONF_OVERRIDES="${KAYOBE_AUTOMATION_CONFIG_PATH}/tempest/tempest.overrides.conf"
+
+if [ ! -z ${KAYOBE_ENVIRONMENT:+x} ]; then
+  KAYOBE_AUTOMATION_TEMPEST_CONF_OVERRIDES="${KAYOBE_AUTOMATION_CONFIG_PATH}/tempest/tempest-${KAYOBE_ENVIRONMENT}-${KAYOBE_AUTOMATION_TEMPEST_LOADLIST:-}.overrides.conf"
+
+  # Check if loadlist specific overrides exist, if not fallback to environment overrides.
+  if [ ! -e "${KAYOBE_AUTOMATION_TEMPEST_CONF_OVERRIDES}" ]; then
+      KAYOBE_AUTOMATION_TEMPEST_CONF_OVERRIDES="${KAYOBE_AUTOMATION_CONFIG_PATH}/tempest/tempest-${KAYOBE_ENVIRONMENT}.overrides.conf"
```
Comment (jovial) [inline .automation.conf/config.sh:15]: Just realised we should probably check `${KAYOBE_AUTOMATION_CONFIG_PATH}/tempest/tempest-${KAYOBE_ENVIRONMENT}.overrides.conf`exists before setting KAYOBE_AUTOMATION_TEMPEST_CONF_OVERRIDES. That way it falls back to `${KAYOBE_AUTOMATION_CONFIG_PATH}/tempest/tempest.overrides.conf`

Comment (markgoddard) [reply .automation.conf/config.sh:15]: Done

**Review by jovial — APPROVED** [review]
New changes look good to me.

**Pull request merged**